File: Caesar/ModuloArithmetic.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

[assembly: InternalsVisibleTo("CaesarTest")]

namespace Caesar
{
    class ModuloArithmetic
    {
        private readonly int mod;

        public ModuloArithmetic(int mod)
        {
            this.mod = mod;
        }

        public int Congruent(int v)
        {
            return (v >= 0) ? v % mod : Congruent(v + mod);
        }

        public int Inverse(int x) 
        {
            if (x < 0) return this.Inverse(this.Congruent(x));

            var r = Egcd(26, x);
            if (r.Item1 > 1) throw new ArgumentException(string.Format("{0} has no inverse in {1} (gcd({0}, {1}) = {2}", x, mod, r.Item1));
            return Egcd(26, x).Item3;
        }

        public static Tuple<int, int, int> Egcd(int a, int b) 
        {
            if (b == 0) return Tuple.Create(a, 1, 0);
            var dstOld = Egcd(b, a % b);
            var dst = Tuple.Create(dstOld.Item1, dstOld.Item3, dstOld.Item2 - (a / b) * dstOld.Item3);

            return dst;
        } 
    }
}

File: Caesar.F/ModuloArithmetic.fs
namespace Caesar.F

type ModuloArithmetic(modulo: int) = 
    static member Egcd(a: int, b: int) : int * int * int =
        if b = 0 then (a, 1, 0)
        else 
            let (d', s', t') = ModuloArithmetic.Egcd(b, a % b)
            (d', t', s' - (a / b) * t')
        
    member this.Congruent(v: int) =
        if v < 0 then this.Congruent(v + modulo);
        else v % modulo

    member this.Inverse(x: int) = 
        if x < 0 then this.Inverse(this.Congruent(x))
        else
            let (ggt, t, s) = ModuloArithmetic.Egcd(modulo, x)
            if ggt <> 1 then raise (System.ArgumentException("ggt != 1"))
            else s